### ISSUE TITLE
Tracking Issue for `once_cell`

### ISSUE THREAD

**Issue by KodrAus** [rust-lang/rust#74465]
This is a tracking issue for the RFC "standard lazy types" (rust-lang/rfcs#2788).
The feature gate for the issue is `#![feature(once_cell)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Complete the RFC process over at https://github.com/rust-lang/rfcs/pull/2788
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

Inlined from #72414:

- [ ] Naming. I'm ok to just roll with the `Sync` prefix like `SyncLazy` for now, but [have a personal preference for `Atomic`](https://github.com/rust-lang/rfcs/pull/2788#issuecomment-574466983) like `AtomicLazy`.
- [x] [Poisoning](https://github.com/rust-lang/rfcs/pull/2788#discussion_r366725768). It seems like there's [some regret around poisoning in other `std::sync` types that we might want to just avoid upfront for `std::lazy`, especially if that would align with a future `std::mutex` that doesn't poison](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/parking_lot.3A.3AMutex.20in.20std/near/190331199). Personally, if we're adding these types to `std::lazy` instead of `std::sync`, I'd be on-board with not worrying about poisoning in `std::lazy`, and potentially deprecating `std::sync::Once` and `lazy_static` in favour of `std::lazy` down the track if it's possible, rather than attempting to replicate their behavior. cc @Amanieu @sfackler.
- [ ] [Consider making`SyncOnceCell::get` blocking](https://github.com/matklad/once_cell/pull/92). There doesn't seem to be consensus in the linked PR on whether or not that's strictly better than the non-blocking variant.

### Implementation history

- https://github.com/rust-lang/rust/pull/68198 (closed in favor of #72414)
- https://github.com/rust-lang/rust/pull/72414

**Comment by matklad** [comment]
Let's cross-out the "should get be blocking?" concern. I decided against this for once_cell, for the following reasons:

* it's makes Clone, Eq, Debug blocking, which is surprising
* the original issue that prompted this question used Lazy, and Lazy is immune from this issue, as it always uses blocking `get_or_init`.

**Comment by matklad** [comment]
Added two more open questions from the RFC.

**Comment by matklad** [comment]
I've added a summary of proposed API to the issue description. 

I wonder if makes sense for @rust-lang/libs to do a sort of "API review" here: this is a pretty big chunk of API, and we tried to avoid bike shedding on the RFC.

**Comment by matklad** [comment]
Here's an interesting use-case for non-blocking subset of OnceCell -- building cyclic data structures: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4eceeefc224cdcc719962a9a0e1f72fc

**Comment by withoutboats** [comment]
I strongly expect a method called `get` to be nonblocking. I am softly in favor of adding a `wait` API that blocks, but would prefer that it be added in a separate feature later based on demand.

**Comment by matklad** [comment]
Yeah, to be clear, there's a consensus that `get` should be non-blocking, the question is resolved. What is not completely solved in my mind, is where we should have `core::lazy::SyncOnceCell`. That's possible in theory (by only providing `get` and `set` methods), but would be hacky to implement, and of questionable usefulness. The above example is a new use-case for that thing.

**Comment by m-ou-se** [comment]
> Naming. I'm ok to just roll with the Sync prefix like SyncLazy for now, but have a personal preference for Atomic like AtomicLazy.

I don't think `Atomic` would be the right word for these. Rust's `Atomic` types and operations (including `Arc`) never block and never involve the operating system's scheduler (they're all defined in `core` or `alloc`, not std). They're all directly based on the basic atomic operations supported by the processor architecture itself.

I'd expect something that's named AtomicLazy/AtomicOnceCell to do the same. And that's something that already exists as another valid strategy for certain Lazy/OnceCell-like types: Instead of blocking all but one thread when multiple threads encounter an 'empty' cell, it wouldn't block but run the initialization function on each of these threads. The first thread to finish atomically stores its initialized value in the cell, and the others simply `drop()` the value they created.

The `std` does something similar in a few places (although not wrapped in a type or publicly exposed). For example, here:

https://github.com/rust-lang/rust/blob/a835b483fe0418b48ca44afb65cd0dd6bad4eb9b/library/std/src/sys/windows/compat.rs#L65-L67

And here:

https://github.com/rust-lang/rust/blob/a835b483fe0418b48ca44afb65cd0dd6bad4eb9b/library/std/src/sys/windows/mutex.rs#L119-L133

So, since this Lazy/OnceCell implementation does block (such that the initialization function can be `FnOnce` and the type doesn't have to fit in an atomic), and an alternative purely atomic strategy does exist, I'd really avoid using the word 'atomic' in the name here.

**Comment by matklad** [comment]
I've added non-blocking flavors of the primitives to the once_cell crate: https://docs.rs/once_cell/1.5.1/once_cell/race/index.html. They are restricted (can be provided only for atomic types), but are compatible with no_std.

It seems to me that "first one wins" is a better semantics if you can't block, so I am going to resolve `Sync no_std subset` like this:

* only `std` supports sync module, as it requires synchronization (`std::thread`)
* if you need something like OnceCell in no-std, your choices are
  * use `race` module from `once_cell` with different API, which might or might not be upifted to std some day
  * use some version based on spin locks (this risks @matklad crashing into issue tracker of your project with explanation of how [pure spin locks are almost always wrong](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html)). 

I've ticked this question's box.

**Comment by m-ou-se** [comment]
It's a bit of a shame that `Lazy` uses a `fn() -> T` by default. With that type, it needlessly stores a function pointer even if it is constant. Would it require big language changes to make it work without storing a function pointer (so, a closure as ZST), while still being as easy to use? Maybe if captureless closures would implement some kind of `const Default`? And some way to not have to name the full type in statics. That's probably not going to happen very soon, but it'd be a shame if this becomes possible and we can't improve `Lazy` because the `fn() -> T` version was already stabilized. Is there another way to do this?

**Comment by phil-opp** [comment]
@matklad 

> They are restricted (can be provided only for atomic types), but are compatible with no_std.

This seems like a very major restriction, which rules out most use cases of `SyncLazy`/`SyncOnceCell`. So I don't think that this really resolves the sync `no_std` use case.

>  - use some version based on spin locks (this risks @matklad crashing into issue tracker of your project with explanation of how [pure spin locks are almost always wrong](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html)).

I agree that spinlocks have their problems, but they're still better than using `static mut` instead. I understand that we don't want to hardcode `SyncLazy`/`SyncOnceCell` to use a deadlock-prone spinlock on `no_std`, but maybe it's possible to let the user supply their own implementation of a Mutex/Once primitive?

This could be implemented using a second generic argument on the `Sync*` types (or maybe even on the `Mutex`/`Once` types). This way, users could specify how the synchronization should happen based on their application. A single-threaded embedded application could just disable interrupts for the critical section, a toy OS kernel could use a spinlock, and projects with their own threading system could supply a "proper" synchronization primitive. Maybe I'm missing something, but this seems like a good solution to me.

**Comment by m-ou-se** [comment]
Some thoughts about `&mut self` functions on (Sync)OnceCell:

These types have both `&mut self` and `&self` functions, but the `&mut` interface seems somewhat incomplete, and it's a bit tricky to pick names for overlapping functionality. For example, `take` can only be done with unique access, so `fn take(&mut self) -> Option<T>` makes sense. But `set` can be done on an empty cell through a shared reference, or on a cell in any state through an unique reference. So both `fn set(&self, value: T) -> Result<&T, T>;` (like Cell::set) and  `fn set(&mut self, value: T) -> &mut T;` (like Option::insert) would make sense.

Maybe if the `get_or_insert`/`get_or_insert_with` pair already provides a 'one time set' functionality, `set` (or `insert`?) should be the `&mut self` version instead?

**Comment by matklad** [comment]
**Unresolved question: method naming**

Currently, we have `get_or_init` and `get_or_try_init`. Are those good names? Here are some alternatives (see also https://github.com/rust-lang/rust/pull/78943)

1) `get_or_init`, `get_or_try_init`
2) `get_or_insert_with`, `try_get_or_insert_with`
3) `get_with`, `try_get_with`

**1**. Pro: Status Quo, name specific to `OnceCell` (you see `x.get_or_init`, you know `x` is one cell). Con: doesn't feel like it perfectly fits with other std names.
**2**. Pro: matches `Option::get_or_inser_with` exactly. Con: for OnceCell, unlike `Option`, this is the core API. It's a shame that its a mouthful.
**3**. Pro: short, matches std conventions. Con: `_with` without `or` suggest that the closure will be always called, but it's not really the case. 

I've though about this, and I think I actually like **3** most. It's *Con* seems like a *Pro* to me. In the typical use-case, you only use `_with` methods:

```rust
impl Spam {
  fn get_eggs(&self) -> &Eggs {
    self.eggs.get_with(|| Eggs::cook())
  }
}
```

So, the closure is sort-of *always* called, it's just cached. Not sure if I my explanation makes sense, but I do feel that this is different from, eg, `Entry::or_insert_with`.

**Comment by matklad** [comment]
@phil-opp: I think it is rather certain that, even if std provides a subset of OnceCell for no_std, it will be non-blocking subset (`set` and `get`).

It certainly is possible to use spinlocks, or make sync::OnceCell parametric (compile-time or run-time) over blocking primitives. I am pretty sure that should be left for crates.io crate though. 

I feel one important criterion for inclusion in std is "design space has a solution with a single canonical API". OnceCell API seem canonical. If we add paramters, the design space inflates. Even if some solution would be better, it won't be obviously canonical, and would be better left to crates.io.

**Comment by matklad** [comment]
>It's a bit of a shame that Lazy uses a fn() -> T by default. 

@m-ou-se yeah, totally agree that this is a hack and feels like a hack. It works well enough in practice, but there's one gotcha: specifying type for a local lazy does not work:

```rust
let x = 92;
let works1: = Lazy::new(|| x.to_string());
let broken: Lazy<String> = Lazy::new(|| x.to_string());
let works2: Lazy<String, _> =  Lazy::new(|| x.to_string());
```

The `broken` variant is something that people occasionally write, and it fails with a somewhat confusing error. If we remove the default type, it will still be broken, but folks won't have intuition that "one parameter should be enough".


One easy way out here is to stabilize only `OnceCell`, and punt on `Lazy` for the time being. `OnceCell` contains all the tricky bit, and `Lazy` is just some syntactic sugar. For me (and probably for some, but not all, other folks) writing

```rust
fn global_state() -> &'static GlobalState {
  static INSTANCE: SyncOnceCell<GlobalState> = SyncOnceCell::new();
  INSTANCE.get_or_init(GlobalState::default)
}
```

doesn't feel like a deal breaker.I'd prefer that to pulling a 3rd party dep (lazy_staic or once_cell).


That said, I think `Lazy`'s hack is worth stabilizing. Even if in the future we'll be able to write:

```rust
static GLOBAL_STATE: Lazy<GlobalState, _> = Lazy::new(GlobalState::default);
```

I don't see a lot of practical problems with 

```rust
static GLOBAL_STATE: Lazy<GlobalState> = Lazy::new(GlobalState::default);
```

working as well.

**Comment by nwn** [comment]
> 
> **Unresolved question: method naming**
> 
>     1. `get_or_init`, `get_or_try_init`
> 
>     2. `get_or_insert_with`, `try_get_or_insert_with`
> 
>     3. `get_with`, `try_get_with`
> 

I think 1 is the most appropriate. The `init` terminology makes more sense than `insert` in the context of a once cell. Depending on whether we expose a direct value initializer, it may be more consistent to add `_with` to these methods, though.

> I've though more about this, and I think I actually like 3 most. It's Con seems like a Pro to me. In the typical use-case, you only use _with methods:
> 
> [...]
>
> So, the closure is sort-of always called, it's just cached. Not sure if I my explanation makes sense, but I do feel that this is different from, eg, Entry::or_insert_with.

This doesn't seem very intuitive to me and isn't always true when there are multiple points of initialization. For example, consider:
``` Rust
impl Spam {
    fn get_eggs(&self, cooked: bool) -> &Eggs {
        if cooked {
            self.eggs.set(Eggs::cook());
        }
        self.eggs.get_with(|| Eggs::raw())
    }
}
```
In this case, the closure may not run and in fact a different value has been cached. I think `get_or_init_with` would make this case more clear.

**Comment by raphaelcohn** [comment]
Something I've recently got bitten by is the need to manage which memory allocator a memory uses. I've been workign wit ha design that has a different global memory allocator when running threads or coroutines (so restricting a coroutine to a maximum amount of memory). This could be thought of as a bit of a hack; one of the long-term design decisions of early Rust that still bites is not making the memory allocator type explicit in the standard collections.

With a lazy, the challenge becomes ensuring that they're all allocated using the same memory allocator.

**Comment by KodrAus** [comment]
On the naming and organization question, I've found myself coming back to:

```rust
mod core {
    pub mod cell {
        pub struct OnceCell<T> {}
        pub struct LazyCell<T, F = fn() -> T> {}
    }
}

mod std {
    pub mod lazy {
        pub struct Once<T> {}
        pub struct Lazy<T, F = fn() -> T> {}
    }
}
```

Based on a few observations:

- In my own usage I haven't need the "unsync" versions of `OnceCell` or `Lazy` yet. Just the sync ones, so giving them the canonical name seems preferable to me.
- In wanting to split up the `std::sync` module I find myself wanting to deprecate `sync::Once` in favor of `lazy::Once<()>`, where the poisoning behavior of `sync::Once` could be layered in through something like `lazy::Once<Poison<()>>`.
- `cell` is the module I already go to to find non-`Sync` shared mutable containers

We looked at this API briefly in #68198 but didn't want to hold up landing _something_, so didn't really spend much time on it.

**Comment by matklad** [comment]
I find "give sync versions the canonical name based on usage frequency" to be a strong and compelling argument. I like `std::lazy::{Once, Lazy}` and I think we should do that, provided that we do want to deprecate `std::sync::Once` eventually. 

I find `cell::{OnceCell, LazyCell}` relatively less compelling, using `Cell` suffix to mean `!Sync` feels like a hack. But it's not like we have an obviously better alternative. 

We *can* consider providing only std sync versions, but I'd be against that. I think it's important not to pessimize single-threaded use-cases. I find that single-threaded shared-nothing architectures *might* be the way to get the most out of many-core machines, and it's Rust unique advantage to check shared-nothingness at compile time. 

Does anybody want to send a PR with the reorg according to https://github.com/rust-lang/rust/issues/74465#issuecomment-763993225? It might make sense to leave deprecated aliases in place, to not make the life of nightly users more complicated.

**Comment by Zenithsiz** [comment]
Is implementing `DerefMut` or having some `get_mut_or_init` for non-sync `Lazy` a planned feature?

My use case is conditionally locking a mutex depending on a branch,
making it so that it doesn't get locked if never dereferenced.
Something along the lines of this:
```rust
// `value` is `Mutex<T>`.
let mut value = Lazy::new(|| value.lock().expect("Poisoned"))
match other_value {
  0 => *value += 1,
  2 => *value -= 1,
  _ => (),
}
```
so that I don't have to repeat `value.lock().expect("Poisoned")` in each branch, nor introduce a local within each match arm while not locking the mutex if we get to the `_` branch.

**Comment by rcls** [comment]
What's the intended way to return a Lazy from a function (or store in a data-structure) without exposing the second type parameter?
The best I could come up with is the fairly fugly:
```
#![feature(once_cell)]
use std::lazy::Lazy;
use std::ops::Deref;
pub fn strint(x: u64) -> impl Deref<Target = String> {
    Lazy::new(move || x.to_string())
}
```
Even just implementing `Into` appropriately would make this saner?

**Comment by matklad** [comment]
@rcls in those cases, it would probably be prudent to build off `OnceCell` directly, as that gives your more flexibility with respect to lifetimes. For your example, I'd do:

```rust
pub struct LazyString {
  x: u64,
  cell: OnceCell<String>,
}

impl LazyString {
  pub const fn new(x: u64) -> LazyString { LazyString { x, cell: OnceCell::new() } } 
}

impl Deref for LazyString {
  type Target = String;
  fn deref(&self) -> &String { self.cell.get_or_init(|| self.x.to_string()) }
}
```

**Comment by SimonSapin** [comment]
Filed a minor docs bug: https://github.com/rust-lang/rust/issues/85716

**Comment by anka-213** [comment]
Is there any specific reason why something like `into_inner` doesn't/shouldn't exist for `Lazy` and `SyncLazy`, as it does for `OnceCell` and `SyncOnceCell`?

I believe something like this would work
```
    pub fn into_inner(self) -> T {
        match self.cell.into_inner() {
            Some(x) => x,
            None => match self.init.take() {
                Some(f) => f(),
                None => panic!("`Lazy` instance has previously been poisoned"),
            },
        }
    }
```

**Comment by matklad** [comment]
No specific reason. once_cell has into_value for that: https://docs.rs/once_cell/1.8.0/once_cell/sync/struct.Lazy.html#method.into_value, it could be added here as well.

**Comment by yaahc** [comment]
> **Unresolved question: method naming**
> 
> Currently, we have `get_or_init` and `get_or_try_init`. Are those good names? Here are some alternatives (see also #78943)
> 
>     1. `get_or_init`, `get_or_try_init`
> 
>     2. `get_or_insert_with`, `try_get_or_insert_with`
> 
>     3. `get_with`, `try_get_with`
> 
> 
> **1**. Pro: Status Quo, name specific to `OnceCell` (you see `x.get_or_init`, you know `x` is one cell). Con: doesn't feel like it perfectly fits with other std names.
> **2**. Pro: matches `Option::get_or_inser_with` exactly. Con: for OnceCell, unlike `Option`, this is the core API. It's a shame that its a mouthful.
> **3**. Pro: short, matches std conventions. Con: `_with` without `or` suggest that the closure will be always called, but it's not really the case.
> 
> I've though more about this, and I think I actually like **3** most. It's _Con_ seems like a _Pro_ to me. In the typical use-case, you only use `_with` methods:
> 
> ```rust
> impl Spam {
>   fn get_eggs(&self) -> &Eggs {
>     self.eggs.get_with(|| Eggs::cook())
>   }
> }
> ```
> 
> So, the closure is sort-of _always_ called, it's just cached. Not sure if I my explanation makes sense, but I do feel that this is different from, eg, `Entry::or_insert_with`.

I definitely favor 3 as well. IMO the `OnceCell` itself implies the fact that it's only called once, so `OnceCell::get_with` still communicates the same semantics that `*_or_*` do on other types.

**Comment by yaahc** [comment]
Summarizing some of the backlog related to open issues:

* Mara: Atomic doesn't feel appropriate. ([source](https://github.com/rust-lang/rust/issues/74465#issuecomment-703295845))
* Mara: Can we change `Lazy` so it doesn't unconditionally store a function pointer? ([source](https://github.com/rust-lang/rust/issues/74465#issuecomment-725462718))
* phil-opp: Could the no-std usecase for `Lazy` be improved by adding a generic parameter for the mutex type? ([source](https://github.com/rust-lang/rust/issues/74465#issuecomment-725472632))
* Mara: is `OnceCell::set` equivalent to `Cell::set` or to `Option::insert` and based on the analogous interface, should it take `&mut self` or `&self`? ([source](https://github.com/rust-lang/rust/issues/74465#issuecomment-725473912))

Didn't get as far through the backlog as I wanted so some of these may already be resolved by later comments.

**Comment by programmerjake** [comment]
> So both `fn set(&self, value: T) -> Result<&T, T>;` (like Cell::set) and `fn set(&mut self, value: T) -> &mut T;` (like Option::insert) would make sense.

imho we should have `set` take `&self`, the `&mut self` variant can instead be spelled `*self = OnceCell::from(value)`.

**Comment by WaffleLapkin** [comment]
@matklad Is the rename/move proposed in https://github.com/rust-lang/rust/issues/74465#issuecomment-763993225 still considered? https://github.com/rust-lang/rust/issues/74465#issuecomment-783294558 asked if someone could make a PR and I don't see anyone responding to this. I'd like to make a PR, if it's still an option

----

Currently, the `set` methods look like this

```rust
pub fn set(&self, value: T) -> Result<(), T> {}
```

I'm a bit surprised that it doesn't return a reference to the value (either old or just set) like `Option::insert`, `{HashMap,BTreeMap}::try_insert`. 

Maybe we could change `set` to one of the following signatures? 

1)
   ```rust
   pub fn set(&self, value: T) -> Result<&T, (&T, T)> {}
   ```
2)
   ```rust
   // maybe poorly named, just an example
   pub struct AlreadySet<'a, T> { 
       pub already_set_to: &'a T,
       pub value: T,
   }
   pub fn set(&self, value: T) -> Result<&'a T, AlreadySet<'_, T>> {}
   ```
3)
   ```rust
   pub fn set(&self, value: T) -> (&'a T, Result<(), T>) {}
   ```
4)
   ```rust
   // maybe poorly named, just an example
   pub enum SetResult<'a, T> { Set(&'a T), WasSet(&'a T, T) }
   pub fn set(&self, value: T) -> (&'a T, Result<(), T>) {}
   ``` 

That would allow using the value immediately after setting it without a need to `unwrap`:
```rust
// Currently unwrap is needed
cell.set(value);
let r = cell.get().unwrap();
```

**Comment by KodrAus** [comment]
Things get buried in discussion threads so I think it’s worth restating that I think any stabilization plan for what’s currently called `SyncOnceCell` that doesn’t consider deprecation of the existing `sync::Once` as an end state is going to leave us in a confusing place where we have multiple APIs for doing the same thing with their differences buried mostly in trivia. I think that should factor into naming and organization.

**Comment by SimonSapin** [comment]
Can we do library API deprecation such that usage only emits warnings in a new edition?

`sync::Once` has been the only way to do some things for a long time. Directing new code to use `SyncOnceCell` instead is good, but pushing existing code to migrate can feel like churn for not much benefit.

**Comment by KodrAus** [comment]
> pushing existing code to migrate can feel like churn for not much benefit.

That’s a fair point. Maybe for a start we could just do a “soft deprecation” where the docs for `sync::Once` suggest you use `SyncOnceCell` instead.

**Comment by vultix** [comment]
Tokio provides a `Future` based [OnceCell](https://docs.rs/tokio/1.12.0/tokio/sync/struct.OnceCell.html)  type.  This might be entirely out of scope, but is there any way the standard `OnceCell` and `Lazy` types might support both synchronous and asynchronous operations?

I imagine in the future we could add a signature similar to this:
```rust
pub async fn get_or_init_async<F, Fut>(&self, f: F) -> &T
    where
        F: FnOnce() -> Fut,
        Fut: Future<Output = T>,
{}
```

**Comment by Zenithsiz** [comment]
@vultix Wouldn't `async { value.get_or_init(f().await) }` be equivalent? Both would have type `impl Future<Item = &T>`.

**Comment by CoffeeSip** [comment]
Is there any update on when this feature will be pushed into stable?

**Comment by cdecompilador** [comment]
It would be great to have the option to reset the Lazy to its initial state (evaluating the initialization again) and (deinitialize) for example, here is a use case:
```rust
const FILE: Lazy<&'static str> = Lazy::new(|| {
        if cfg!(debug_assertions) {
            let buf = std::fs::read_to_string("test.txt").unwrap();
            Box::leak(buf.into_boxed_str())
        } else {
            include_str!("test.txt")
        }
});
```

Later in the code
```rust
// Some lets say http server loop
loop {
    // ...
    if cfg!(debug_assertions) {
        FILE.reset();
    }
    // ...
}
```

This would allow in debug build to have live file reloading for `test.txt` while in release build the file is completely static.

**Comment by EFanZh** [comment]
@cdecompilador The reference a global `SyncLazy` gives has `'static` lifetime, which should always be valid, so you should never destroy a `SyncLazy` value after creation. For example:

```rust
#![feature(once_cell)]

use std::lazy::SyncLazy;

static X: SyncLazy<i32> = SyncLazy::new(|| 7);
static Y: SyncLazy<&i32> = SyncLazy::new(|| &X);

fn main() {
    dbg!(*Y);
}
```

You can’t reset `X` because `Y` should always be valid.

**Comment by GutsTang** [comment]
Is there any opportunity to implement the `Copy` trait for `Lazy` ？
When I tried to use `Lazy` in a `const` variable, the closure inside was called multiple times.
This behavior is extremely counter-intuitive.

```rust
#![feature(once_cell)]
use rand::Rng; // 0.8.4
use std::lazy::Lazy;

const CONST_LAZY: Lazy<i32> = Lazy::new(|| rand::thread_rng().gen::<i32>());

fn main() {
    let local_lazy: Lazy<i32> = Lazy::new(|| rand::thread_rng().gen::<i32>());
    println!("{}", *local_lazy); // -1475423855
    println!("{}", *local_lazy); // -1475423855

    println!("{}", *CONST_LAZY); // 1975106939
    println!("{}", *CONST_LAZY); // -1848043613
}
```

**Comment by jRimbault** [comment]
@GutsTang this is expected behavior, what you want is `static` not `const`. `const` means each instance is the same _compile time_ value, think of it almost like `#define N 1`.

What you wrote would be equivalent to :

```rust
println!("{}", *Lazy::new(|| rand::thread_rng().gen::<i32>()));
println!("{}", *Lazy::new(|| rand::thread_rng().gen::<i32>()));
```

**Comment by GutsTang** [comment]
> Thanks. I'm always confused about the `const` variable in Rust.

**Comment by ayosec** [comment]
> this is expected behavior, what you want is `static`. `const` means each instance is the same _compile time_ value, think of it almost like `#define N 1`.

Maybe it would be useful to have a Clippy check for this. I guess that something like `std::lazy::Lazy` should never be stored in a `const` value.

**Comment by Nemo157** [comment]
There is an existing default-warn clippy lint for this:

```
warning: a `const` item should never be interior mutable
 --> src/lib.rs:2:1
  |
2 | pub const LAZY: std::lazy::Lazy<i32> = std::lazy::Lazy::new(|| 5);
  | -----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | make this a static item (maybe with lazy_static)
  |
  = note: `#[warn(clippy::declare_interior_mutable_const)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const
```

**Comment by TopologicallySpeaking** [comment]
> I'm a bit surprised that it doesn't return a reference to the value (either old or just set) like `Option::insert`, `{HashMap,BTreeMap}::try_insert`.
> 
> Maybe we could change `set` to one of the following signatures?
> 
>     1. ```rust
>        pub fn set(&self, value: T) -> Result<&T, (&T, T)> {}
>        ```
> 
>     2. ```rust
>        // maybe poorly named, just an example
>        pub struct AlreadySet<'a, T> { 
>            pub already_set_to: &'a T,
>            pub value: T,
>        }
>        pub fn set(&self, value: T) -> Result<&'a T, AlreadySet<'_, T>> {}
>        ```
> 
>     3. ```rust
>        pub fn set(&self, value: T) -> (&'a T, Result<(), T>) {}
>        ```
> 
>     4. ```rust
>        // maybe poorly named, just an example
>        pub enum SetResult<'a, T> { Set(&'a T), WasSet(&'a T, T) }
>        pub fn set(&self, value: T) -> (&'a T, Result<(), T>) {}
>        ```
> 
> 
> That would allow using the value immediately after setting it without a need to `unwrap`:
> 
> ```rust
> // Currently unwrap is needed
> cell.set(value);
> let r = cell.get().unwrap();
> ```

I strongly agree. I'm currently writing a builder pattern, the builder has a number of items which need to be set one after another, each one depends on the last, and there are multiple references to the builder so I can't mutably borrow it. That's exactly what `OnceCell` is designed for. I should be able to do something roughly like this:

```rust
fn build(&self) -> BuiltType {
    let item1 = self
        .item1
        .set(Item1::new(self.info))
        .unwrap();

    self.item2
        .set(Item2::new(item1))
        .unwrap();

    self.finish()
}
```

Option 1 and 2 are the only ones you listed which would work for that purpose. I'd prefer the second one, as it's more explicit about the semantics.

**Comment by zseri** [comment]
@TopologicallySpeaking huh, but I think you could also work with the third option, e.g.
```rust
fn build(&self) -> BuiltType {
    let (item1, y) = self
        .item1
        .set(Item1::new(self.info));
    y.unwrap();

    self.item2
        .set(Item2::new(item1))
        .unwrap();

    self.finish()
}
```

**Comment by DefinitelyNotRobot** [comment]
Not sure if this has been discussed, but I think `OnceCell` should potentially belong in `std::cell` rather than `std::lazy`, since its use-case can be more general than for laziness. Similarly, perhaps `SyncOnceCell` should belong in either `std::cell` or `std::sync`.

**Comment by zseri** [comment]
@DefinitelyNotRobot If `OnceCell` belongs in `std::cell`, then `SyncOnceCell` belongs in `std::sync`, like `Rc`/`Arc`.

**Comment by andylokandy** [comment]
Is `OnceCell` safe in static considering multiple threads will try to initialize it? I've seen some usage like that, for example:
https://github.com/metrics-rs/quanta/blob/fbf383a33d7836d7303dd3aa8d9627e17cd613da/src/lib.rs#L170

**Comment by SimonSapin** [comment]
@andylokandy There are two variants of `OnceCell`. One of them implements the `Sync` trait, which means it is safe to access from multiple threads. (The language won’t let you make a `static` item of a `!Sync` type.) Both variants exist because this extra synchronization has some cost. In your example, not the import `use once_cell::sync::OnceCell;` instead of `once_cell::unsync::OnceCell`.

**Comment by Person-93** [comment]
Proposed addition to the API for  `OnceCell`: initialization that returns a mutable reference.
```rust
// regular initialization
let mut cell = OnceCell::new();
let n: &mut i32 = cell.get_mut_with(|| 42);

// fallible initialization
let mut cell = OnceCell::new();
let n: &mut i32 = cell.try_get_mut_with(|| {
 if all_is_well {
    Ok(42)
  } else {
    Err("oh no!")
  }
})?;
```
NOTE: I'm not sure about the method names.

**Comment by SimonSapin** [comment]
If you have a `&mut OnceCell<T>` during initialization why would you need `OnceCell` at all? If you only have `&OnceCell<T>`, any method that returns `&mut T` from that would be unsound. Only `&T` ever being accessible through `&OnceCell<T>` is a basic principle of `OnceCell`.

**Comment by Person-93** [comment]
> If you have a `&mut OnceCell<T>` during initialization why would you need `OnceCell` at all? If you only have `&OnceCell<T>`, any method that returns `&mut T` from that would be unsound. Only `&T` ever being accessible through `&OnceCell<T>` is a basic principle of `OnceCell`.

My use case is where initialization can be expensive so it should only happen once but updates can happen often and are usually pretty cheap.

**Comment by SimonSapin** [comment]
If there’s shared access and there can be any update after initialization, `OnceCell` does not work. That’s why it’s called "once". You need a `RefCell` or a `Mutex` or a `RwLock` instead.

Allowing mutation after initialization in `OnceCell` would make it unsound. For example:

```rust
let mut cell = OnceCell::new();
let a: &mut Option<String> = cell.get_mut_with(|| Some(String::from("foo")));
let b: &Option<String> = cell.get(); // `&T` aliases `&mut T`
let c: &str = b.as_deref().unwrap()
*a = None; // String is deallocated
println!("{c}") // Use-after-free

**Comment by Shadlock0133** [comment]
It is possible to implement it completely outside of `std`:

```rust
fn get_mut_with<T, F: FnOnce() -> T>(cell: &mut OnceCell<T>, f: F) -> &mut T {
    cell.get_or_init(f);
    cell.get_mut().unwrap()
}
```

But I still don't know if it's that useful.

**Comment by Person-93** [comment]
> If there’s shared access and there can be any update after initialization, `OnceCell` does not work. That’s why it’s called "once". You need a `RefCell` or a `Mutex` or a `RwLock` instead.
> 
> Allowing mutation after initialization in `OnceCell` would make it unsound. For example:
> 
> ```rust
> let mut cell = OnceCell::new();
> let a: &mut Option<String> = cell.get_mut_with(|| Some(String::from("foo")));
> let b: &Option<String> = cell.get(); // `&T` aliases `&mut T`
> let c: &str = b.as_deref().unwrap()
> *a = None; // String is deallocated
> println!("{c}") // Use-after-free
> ```

The call to `cell.get()` to assign to `b` wouldn't work because `cell` is mutably borrowed.

**Comment by SimonSapin** [comment]
So `get_mut_with` would take `&mut self`? But if you can afford exclusive access for initialization, why use a cell at all?

```rust
// regular initialization
let mut value;
value = 42;
let n: &mut i32 = &mut value;

// fallible initialization
let mut value;
if all_is_well {
    value = 42;
} else {
    return Err("oh no!")
}
let n: &mut i32 = &mut value;

**Comment by Person-93** [comment]
> So `get_mut_with` would take `&mut self`? But if you can afford exclusive access for initialization, why use a cell at all?

If initialization is expensive or not idempotent, I want to ensure that it only happens once.

**Comment by matklad** [comment]
@Person-93 perhaps you are looking for https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.get_or_insert_with?  Otherwise, I'd say https://github.com/rust-lang/rust/issues/74465#issuecomment-1060033519 is the way to go.

As a policy question, I would probably suggest to *not* discuss additional convenience methods in this issue, but rather direct issue requests at once_cell crate.

Here, I feel we can benefit most from focusing on the minimal useful subset we can stabilize **first**. After we have a stable base, it would be much more efficient to open follow-up issues/pr expanding the API surface.

**Comment by matklad** [comment]
I ... have a new naming proposal:

* `OnceCell` for !Sync version
* `OnceLock` for `Sync` version

I think I **really**  like OnceLock:

* It *is* a lock! (blocking inside)
* It make the semantics clear (locking rather than racy publication)
* It "rhymes" with `RwLock`

**Comment by m-ou-se** [comment]
That sounds great. I'm wondering if this means they shouldn't live in the `{core, std}::lazy` module, but instead in `core::cell` and `std::sync`.

**Comment by bstrie** [comment]
Agreed, I like it. (And I am getting to this, just slowly! Someone with more free time is welcome to beat me to it, of course. :) )

I will say that `std::sync` feels like sort of a miscellaneous grab bag, but `std::sync::OnceLock` doesn't feel terrible.

**Comment by cuviper** [comment]
It might be confusing to have `std::sync::{Once, OnceLock}` next to each other, because the difference is not really obvious from their names alone. At the very least, docs should anticipate this confusion and try to clarify their roles -- `OnceLock` is roughly just `(Once, T)`.

**Comment by cuviper** [comment]
A small point against the `OnceLock` name is that you don't have guards like `Mutex` or `RwLock`. In that sense, you never actually "lock" the data inside.

**Comment by cuviper** [comment]
Alternate name idea: `InitLock` -- as in, only initialization is actually locked, and then it has unguarded access.

**Comment by KodrAus** [comment]
@cuviper Another way to look at it is that `Once` is `OnceLock<()>`, which is my personal preference because it means we can deprecate `Once` altogether.

**Comment by SimonSapin** [comment]
If you mean soft-deprecation in docs, why not. But making `Once` emit a deprecation warning to push its existing users to migrate to `OnceLock` would create churn. For what benefit? Since `Once` is `#[stable]` so we can’t actually remove it.

**Comment by yaahc** [comment]
> If you mean soft-deprecation in docs, why not. But making `Once` emit a deprecation warning to push its existing users to migrate to `OnceLock` would create churn. For what benefit? Since `Once` is `#[stable]` so we can’t actually remove it.

I know we can't really remove it from the compiled library, but I still do like the idea of eventually removing the ability to call deprecated APIs in newer editions. The benefit in my mind being keeping the language surface area as small as possible. I think there is a significant cost in having multiple ways to do the same thing, if people can use `Once` or `OnceLock<()>` to do the exact same thing, it benefits them for us to push everyone into using the newer more powerful API so you don't end up with people learning old dialects and having to eventually unlearn the old API in favor of the new one.

**Comment by nbdd0121** [comment]
How about making `std::sync::Once` `Once<T = ()>`?

**Comment by cuviper** [comment]
> But between a small amount of diff churn and having to endlessly maintain two ways to do the same thing, I'd choose the diff churn and the smaller library API every time.

We still do have to maintain the old stuff, even if it's inaccessible from newer editions.

**Comment by bjorn3** [comment]
> How about making `std::sync::Once` `Once<T = ()>`?

That is unfortunately a breaking change. Generic param defaults only work in some cases but not others. Don't remember exactly when they do and when they don't work though.

**Comment by SimonSapin** [comment]
If deprecation is tied to an edition and there’s a rustfix migration, that sounds fine. It’s new deprecation warnings in existing code just by upgrading the toolchain that I feel are probably not worth the unexpected churn if it’s only to simplify std learnability.

**Comment by nbdd0121** [comment]
> That is unfortunately a breaking change. Generic param defaults only work in some cases but not others. Don't remember exactly when they do and when they don't work though.

It wouldn't break anything if we just change the struct, but we can't change `new` from `impl Once` to `impl<T> Once<T>` because default type parameters won't affect type param fallback.

**Comment by John-Nagle** [comment]
What's the status on this? I wanted to use it for something that generates strings for internationalization, but it's still unstable. Is this going to become stable soon?

**Comment by matklad** [comment]
This is blocked on implementation work. I feel this can become stable fairly soon, if this issue gets someone who'd champion remaining implementation&stabilization work. So far this didn't happen.

**Comment by djc** [comment]
(But you can use the once_cell crate in the meantime?)

**Comment by John-Nagle** [comment]
> (But you can use the once_cell crate in the mean time?)

True. The use case involves Egui, the menu system often used for games. On every frame, it goes through the items it needs to display. Those are **&str** parameters. If I add internationalization, there's a lookup for every menu item on every frame. Hence the need for some kind of memoization.

**Comment by WaffleLapkin** [comment]
> This is blocked on implementation work.

Can someone outline what implementation work is needed? This issue has quite a few of comments and it's somewhat hard to get what is needed to be done :')

**Comment by matklad** [comment]
I'd start with the following plan, adjusting according to reality: 

1. Do the rename proposed in https://github.com/rust-lang/rust/issues/74465#issuecomment-1095665053, that is

```rust
mod cell {
    OnceCell,
    CellLazy,
}
mod sync {
    OnceLock,
    LockLazy,
}
```

2. Re-read the disucssion here and on the RFC to make sure nothing is missed.

3. Take a critical look at the naming, signatures and semantics of `OnceX` methods, update the summary comment in the tracking issue.

4. Take the minimal orthogonal set of methods of `OnceX`, start FCP to stabilize thouse under `base_once_cell` feature flag

5. After `OnceX` types are stabilized, move forward with stabilizing `Lazy` types (I'd expect that to need a bit more bikeshed
   due to less clear naming and objectonable default parameter trick).

Throughout: keep an eye on what's the current blocked, poke relevant people and otherwise ensure that the work doesn't get stuck.

**Comment by MatrixDev** [comment]
Why `CellLazy` instead of `LazyCell` (same for lock)?
It is harder to pronounce (triple consecutive L) and also all other cells are suffixes instead of prefixes.

**Comment by metasim** [comment]
> Why `CellLazy` instead of `LazyCell` (same for lock)? 

`LazyCell` is also more coherent with `RefCell`....

**Comment by mqudsi** [comment]
From a usability/ergonomics perspective, is there any chance of adding a `set(value: T)` method to `Lazy<T>` that would be used if-and-only-if the lazy value hasn't already been initialized with the constructor-provided lazy init method (panicking/failing otherwise)?

The idea is that you could specify a default value and override it "in-place" rather than needing to use a `OnceCell<T>` and a separate `mut T` instance (or a third, default const `T` instance) to hold the default value with the option for a user/environment/runtime-provided alternative taking its place, while still only initializing once.

e.g. in lieu of the following:

```rust
static DEST_PORT OnceCell<u16> = OnceCell::new();

pub fn main() {
    // This is the default value, used if no alternative is provided
    let mut dest_port: u16 = 1024;
    ...
    if let Some(arg) = std::env::args().skip(1).next() {
        dest_port = arg.parse().expect("Expected a valid u16 port number as the argument");
    }

    DEST_PORT.set(dest_port).unwrap();
}
```

you would be able to do something like this:

```rust
// This is the default value, used if no alternative is provided
static DEST_PORT Lazy<u16> = Lazy::new(|| 1024u16);

pub fn main() {
    if let Some(arg) = std::env::args().skip(1).next() {
        let dest_port = arg.parse().expect("Expected a valid u16 port number as the argument");
        DEST_PORT.set(dest_port).unwrap();
    }

    // At some later point:
    let foo: u16 = *DEST_PORT;
}
```

**Comment by hamza1311** [comment]
> Naming. I'm ok to just roll with the `Sync` prefix like `SyncLazy` for now, but [have a personal preference for `Atomic`](https://github.com/rust-lang/rfcs/pull/2788#issuecomment-574466983) like `AtomicLazy`.

I would vote in favor of `sync` module. `core::lazy` has the unsync implementation and `core::lazy::sync` has the sync ones. This is also consistent with the naming used for `Arc`, `Mutex`, etc -- those are also in a module named `sync`

**Comment by Iron-E** [comment]
@mqudsi you could do this:

```rust
static DEST_PORT OnceCell<u16> = OnceCell::new();

pub fn main() {
    // This is the default value, used if no alternative is provided
    ...
    if let Some(arg) = std::env::args().skip(1).next() {
        DEST_PORT.set(arg.parse().expect("Expected a valid u16 port number as the argument")).unwrap();
    }

    DEST_PORT.get_or_init(|| 1024);
}
```

**Comment by mqudsi** [comment]
@Iron-E yes, it's possible directly with `OnceCell` but you're back to using a helper function to wrap the `get_or_init()` logic, which is what `Lazy<T>` is supposed to help avoid.

**Comment by Iron-E** [comment]
> in practice you'll have a full command line arg handler processing perhaps dozens of switches, with proper error handling and early abort w/ usage information. 

True… but if you have so many arguments it's probably better to use something like [`clap::Parser`](https://docs.rs/clap/latest/clap/trait.Parser.html) anyway. `Lazy::new(|| Args::parse())` could probably do whatever is needed and more.

I don't see any issue with `Lazy::set` fwiw (not that my opinion is that important!), I just wanted to show how one could already accomplish similar effects by moving some code around.

> The solution isn't to move everything into the Lazy closure; you normally want as little in there as you can get away with.

Agree to disagree I suppose. The two main reasons I can see that one would want something to be `Lazy` is that you either:

1. _Might_ eventually need a value that is somewhat expensive to initialize. 
2. Need a `static` that depends on runtime computation (e.g. a `String`).

Putting as much work in the `Lazy` as possible is better for the first case, but is not strictly better for the latter.

**Comment by rtrevinnoc** [comment]
Hello :) I am sorry if this is not the right place to post this, but could you help me figure out what is the current API? Importing Lazy as such seems to not work anymore. For reference, I am using `rustc 1.64.0-nightly`. Thank you very much beforehand

```rust
#![feature(once_cell)]
use std::lazy::Lazy;
```

**Comment by rusty-snake** [comment]
Read the comments above or look at the documentation (it has a search feature).

`std::lazy::Lazy` got renamed and moved to `std::cell::LazyCell`.

**Comment by Iron-E** [comment]
When you search the docs it brings up [the former](https://doc.rust-lang.org/std/lazy/struct.Lazy.html) :)

**Comment by rusty-snake** [comment]
The docs you linked say

> Version 1.62.0 (a8314ef7d 2022-06-27)

If you use a nightly compiler you need nightly docs.

**Comment by rtrevinnoc** [comment]
Thank you very much, I couldn't find it. I'm so sorry for the disturbance, I should've looked at the nightly docs

**Comment by kennystrawnmusic** [comment]
Attempting to `use core::lazy::OnceCell` in a kernel project is throwing an “unresolved import” error at the moment despite the documentation stating that it’s in the core library, forcing me to fall back on the `conquer_once` implementation at the moment — any idea when this will get fixed?

**Comment by mbartlett21** [comment]
@kennystrawnmusic 
If you look a few comments up, you will see that it was renamed to `core::cell::LazyCell`.
Since you are running nightly, you can either look at your local docs or the [nightly docs on doc.rust-lang.org](https://doc.rust-lang.org/nightly/core/cell/struct.LazyCell.html) to find that it is there.

**Comment by not-wlan** [comment]
FYI: SyncLazy can be found in `std::sync::LazyLock` now

**Comment by nrc** [comment]
ISTM that since OnceCell has a take method which resets it back to the initialised state, it can be set multiple times and therefore 'once' in the name is misleading. Am I missing something about the once-ness?

**Comment by matklad** [comment]
I think this is just a special case of the more general phenomenon of "nothing is immutable, b/c you can always overwrite the variable itself". This is true of the `Once` we already have: 

```rust
fn main() {
    let mut x = std::sync::Once::new();
    x.call_once(|| println!("hello"));
    x = std::sync::Once::new();
    x.call_once(|| println!("hello"));
}
```

This is also similar to the docs for `Mutex`, which currently say

>The data can only be accessed through the RAII guards returned from [lock](https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#method.lock) and [try_lock](https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#method.try_lock), which guarantees that the data is only ever accessed when the mutex is locked.

This is the right intuition, but also a lie, because you can do whatever if you have `&mut Mutex`. 

Admittedly, this is a common question which comes up from time to time https://github.com/matklad/once_cell/issues/153. 

My suggestion would be to stick with `Once` terminology, as it builds the right intuition, and put the details to the `&mut` taking methods.

**Comment by mbartlett21** [comment]
Should the `set` method return `Result<&T, T>` rather than `Result<(), T>`?

**Comment by briansmith** [comment]
Somewhere I found mention of deprecation of `Once` if/when `once_cell` becomes stable. `Once` has an important guarantee that the new types don't make: "When this function returns [it] is also guaranteed that any memory writes performed by the executed closure can be reliably observed by other threads at this point (there is a happens-before relation between the closure and code executing after the return)." See https://github.com/matklad/once_cell/issues/83 for more discussion of this.

Basically, it isn't clear if these new types intend to guarantee anything stronger than "Consume" semantics, regardless of whether the present implementation may (or may not) implement stronger semantics.

**Comment by briansmith** [comment]
I have a codebase where Rust code that ensures a global `static mut GLOBAL_STATE: [u32; 4]` array is initialized before calling into C and assembly code code that directly reads `GLOBAL_STATE` on the assumption it has been initialized beforehand. I investigated adapting this codebase to use the new feature. I see this:

```
pub struct OnceLock<T> {
    once: Once,
    // Whether or not the value is initialized is tracked by `state_and_queue`.
    value: UnsafeCell<MaybeUninit<T>>,
```

In order to support this pattern, it would be convenient to make it so a `OnceLock<T>` can be accessed from non-Rust code as though it were a `T`. That is, it would be great to guarantee that the address of the `OnceLock` is equal to the address of its `value`; i.e. make `value` the first field in the structure, make `OnceLock` be `repr "C"` or equivalent, and ensure `value` is `#[repr(transparent)]`.

**Comment by thomcc** [comment]
> Basically, it isn't clear if these new types intend to guarantee anything stronger than "Consume" semantics, regardless of whether the present implementation may (or may not) implement stronger semantics.

I think anything not guaranteeing Acquire/Release semantics would be too much of a footgun (especially for how minor of a performance benefit it would likely be).

If we ever get some way of emulating Consume, user code can implement its own types that provide that instead, similar to how user code that wants a racy or relaxed behavior currently would need to implement that itself.

---

> That is, it would be great to guarantee that the address of the `OnceLock` is equal to the address of its `value`; i.e. make `value` the first field in the structure, make `OnceLock` be `repr "C"` or equivalent, and ensure `value` is `#[repr(transparent)]`.

I'm a lot less sure about this though. We don't currently do this for any stdlib types (aside from trivial cases like NonZeroFoo and such), do we?

**Comment by briansmith** [comment]
> I'm a lot less sure about this though. We don't currently do this for any stdlib types (aside from trivial cases like NonZeroFoo and such), do we?

`UnsafeCell` itself is `repr(transparent)` so it has this property, IIUC.

**Comment by thomcc** [comment]
Do we ever guarantee the representation of something that isn't purely a wrapper? This would need additional state.

**Comment by Stargateur** [comment]
No, `repr(C)` guarantee order, but compatibility only with most used layout of C compilers of the target.

**Comment by thomcc** [comment]
I meant if any existing types in the stdlib did this that aren't purely wrappers around other types.

**Comment by ibraheemdev** [comment]
Probably worth cross-linking between the sync and unsync versions in their docs.

**Comment by tgross35** [comment]
Having some brief discussion with Thom on Zulip https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/moving.20once.20cell.20forward it seems like it makes sense to guarantee acquire/release at this point, and possibly add a way to specify `consume` if there ever is a use. Which there very well may never be.

The three open issues from the top post are:

- Naming (seems fairly set in stone at this point)
- Acq/Rel vs Consume (doesn't seem like there's any reason to spend effort on `Consume` at this point)
- Method naming (also haven't seen much debate here recently)

The RFC is still open, but it's fairly dependent on the discussion at this issue. It probably just needs some minor updates to the current decisions and then can be merged.

Is there anything blocking FCP aside from the RFC?

**Comment by matklad** [comment]
For lazy, there’s https://github.com/rust-lang/rust/pull/103718.

I think we should fcp something here, yeah.

my personal preference would be for FCPing a minimal subset first (only once cell, no lazy), the api surface is large, we might easily overlook some annoying detail if we try to rubber-stump whatever there is now.

**Comment by WhyNotHugo** [comment]
Not sure if this is the right place to point this out: the documentation for [LazyLock](https://doc.rust-lang.org/std/sync/struct.LazyLock.html) reads:

> This type is a thread-safe `Lazy`, and can be used in statics.

However, there's no type called `Lazy`. This should probably refer to `LazyCell` instead?

**Comment by eggyal** [comment]
Could `OnceCell<T>` be `#[repr(transparent)]` with its layout documented to be identical to `T`?  I would like to initialise a `OnceCell` from FFI without calling back into Rust, by writing directly to the memory location.

**Comment by bjorn3** [comment]
`Option<T>` doesn't have a defined layout either except for `Option<&T>`, `Option<NonNullU*>` and the like.

**Comment by eggyal** [comment]
Indeed.  But in those cases it's useful: `Option<&T>` in my case.

**Comment by eggyal** [comment]
Actually, the same applies to `UnsafeCell`, `SyncUnsafeCell` and `Cell`: whilst they are already `#[repr(transparent)]`, their internal layout is undocumented.

**Comment by tgross35** [comment]
I think `repr(transparant)` would be a bit misleading since it imho kind of guides users to assume that `OnceCell<T>` has the same layout as `T`, not `Option<T>` (the three examples mentioned all have the layout of `T`).

The `get` and `get_mut` methods on `OnceCell` can be used to get pointers to the inner value, would that work in your case?

**Comment by briansmith** [comment]
> No, `repr(C)` guarantee order, but compatibility only with most used layout of C compilers of the target.

The C standards require the address of the first field to be the address of the structure, which is why I suggested `#[repr(c)]`, putting the value at the start of the field, and avoiding using any non-transparent wrappers like `Option` around the field.

Anyway, I don't have a strong opinion about whether to do extra work to support the ability of non-Rust code to be able to access the value.

**Comment by tgross35** [comment]
I don't think there's much benefit to providing any sort of guarantee on internal layout - any alternative to `Option` means mimicing its behavior in a separate place, and losing optimizations geared at `Option` (e.g. niches). For any Rust + C project that already has a good reason to use a Rust `OnceCell`, I really think the correct solution is to use`get()`, `get_mut()`, `get_or_init()`, etc. and wrap them in something `extern "C"`, or pass their result to C as applicable. Otherwise, you're just rewriting those exact functions in C

**Comment by SUPERCILEX** [comment]
Can we add `fn into_inner(self) -> Option<T>` to LazyCell? That'd be helpful when doing things only if the LazyCell fired.

**Comment by eggyal** [comment]
> Can we add `fn into_inner(self) -> Option<T>` to LazyCell? That'd be helpful when doing things only if the LazyCell fired.

Wouldn't `fn is_initialized(this: &Self) -> bool` or perhaps `fn get(this: &Self) -> Option<&T>` be more useful, as they don't require taking ownership?  (As with other smart pointers, these are associated functions to avoid conflicts with methods of the inner type `T`).

**Comment by SUPERCILEX** [comment]
Maybe for others, but for my use case I specifically need to take ownership.

**Comment by SUPERCILEX** [comment]
Went ahead and opened a PR: https://github.com/rust-lang/rust/pull/106152

**Comment by SimonSapin** [comment]
If you have ownership is it useful at all to have a `LazyCell` instead of `Option`? Or is there some scenario where you first need to initialize through a shared reference, and later recover full ownership?

**Comment by SUPERCILEX** [comment]
Well yeah but then I have to manage lazy initialization myself.

**Comment by SimonSapin** [comment]
Like this? https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with

**Comment by SUPERCILEX** [comment]
TIL! That's ever so slightly more annoying b/c you have to carry around the option and closure separately (or wrap them in your own type), but I'd be ok with having my PR closed if we think you should use this instead.

**Comment by elichai** [comment]
> I've added non-blocking flavors of the primitives to the once_cell crate: https://docs.rs/once_cell/1.5.1/once_cell/race/index.html. They are restricted (can be provided only for atomic types), but are compatible with no_std.

FWIW, I use `once_cell` a lot for initializing a cryptographic context, and usually, the `racy` option is the one that I actually want (parking a thread can be more expensive than just initializing the context)

This kind of use case might disappear once `const fn` becomes powerful enough that I'll be able to initialize these at compile time

**Comment by ydewit** [comment]
It just realized that in my specific case, one of the values has a positive polarity and the other one negative. So I can use OnceLock as is by simply storing an enum value with a negative and a positive case.

**Comment by kpark-hrp** [comment]
There appears to be a unexpected behavior when using `const` with `Lazy`. Issue: https://github.com/matklad/once_cell/issues/224

With `const` variable, `Lazy` can get evaluated multiple times and return different results.
I believe the expected behavior should be one of the following.
- does not compile
- `Lazy` is evaluated once, and the result is shared.

**Comment by jhpratt** [comment]
That is simply how `const`s work. There is no reason it should not compile.

**Comment by eggyal** [comment]
> With `const` variable, `Lazy` can get evaluated multiple times and return different results.

That's pretty much what I would expect per the documentation of [`const`](https://doc.rust-lang.org/std/keyword.const.html):

> `const` items looks remarkably similar to `static` items, which introduces some confusion as to which one should be used at which times. To put it simply, constants are inlined wherever they’re used, making using them identical to simply replacing the name of the `const` with its value. Static variables, on the other hand, point to a single location in memory, which all accesses share. This means that, unlike with constants, they can’t have destructors, and act as a single value across the entire codebase.

Perhaps you're after a `static` instead?

**Comment by jplatte** [comment]
Note that there is also a clippy lint for `const`s with interior mutability, which should fire if you use `Lazy` in a `const`: https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const

**Comment by mina86** [comment]
Is there a plan or tracking issue for OnceLock::wait?

**Comment by matklad** [comment]
No, but it seems to me that it should be added (under a separate feature gate&tracking issue).

**Comment by tisonkun** [comment]
Can we have a `get_mut_or_init`?

**Comment by SimonSapin** [comment]
`OnceCell::get_mut_or_init` would only be safe if you have `&mut self`. But if you have exclusive access when initializing, do you even need a cell in the first place? Couldn’t you use a plain `Option` and initialize it with [`get_or_insert_with`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with)?

**Comment by cuviper** [comment]
Can't you say the same about `get_mut`? But it may be a conditional situation, where sometimes you have the necessary information to initialize while you have exclusive access, and other times you need to do it later while shared.

**Comment by tisonkun** [comment]
> OnceCell::get_mut_or_init would only be safe if you have &mut self. But if you have exclusive access when initializing, do you even need a cell in the first place? Couldn’t you use a plain Option and initialize it with [get_or_insert_with](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with)?

Because not all accessors are `get_mut_or_init`. Basically, I need a `get_or_init`, but sometimes I want a `mut` reference and am unsure if it's initialized.

You can read this code snippet and see if there is a better solution than providing a `get_mut_or_init`:

```rust
// batches: OnceCell<Vec<RecordBatch>>,

    pub fn mut_batches(&mut self) -> IterMut<'_, RecordBatch> {
        self.batches.get_or_init(|| load_batches(&self.buf));
        // SAFETY - init above
        unsafe { self.batches.get_mut().unwrap_unchecked() }.iter_mut()
    }


    pub fn batches(&self) -> Iter<'_, RecordBatch> {
        self.batches.get_or_init(|| load_batches(&self.buf)).iter()
    }
```

https://github.com/tisonkun/kafka-api/blob/d080ab7e4b57c0ab0182e0b254333f400e616cd2/kafka-api/src/record.rs#L108-L116

**Comment by tisonkun** [comment]
But you're right that without an async calling environment an `Option` + `get_or_insert_with` may work.

**Comment by tisonkun** [comment]
No. I need to guard possible concurrent access to `batches` so I still need a get_mut_or_init

**Comment by sdroege** [comment]
Are there any plans for making `OnceLock::get_unchecked()` (and I guess the mutable version) public? That's currently the only thing preventing me from moving away from the `once_cell` crate.

**Comment by ChayimFriedman2** [comment]
@sdroege You can do `OnceLock::get().unwrap_unchecked().

**Comment by sdroege** [comment]
Thanks, the atomic op and branch in `get()` is indeed optimized away when doing that.
